cmd/dagger/shell.go

```diff
@@ -13,79 +13,71 @@ import (
-	Use:                   "shell",
-	DisableFlagsInUseLine: true,
-	Hidden:                true, // for now, remove once we're ready for primetime
-	RunE: func(cmd *cobra.Command, args []string) error {
-		focus = queryFocus
-		return loadModCmdWrapper(RunShell, "")(cmd, args)
-	},
-}
-
-func init() {
-	rootCmd.AddCommand(shellCmd)
+var (
+	shellIsContainer bool
+	shellEntrypoint  []string
+)
 
-	shellCmd.Flags().StringVar(&queryFile, "doc", "", "document query file")
-	shellCmd.Flags().StringSliceVar(&queryVarsInput, "var", nil, "query variable")
-	shellCmd.Flags().StringVar(&queryVarsJSONInput, "var-json", "", "json query variables (overrides --var)")
-}
+var shellCmd = &FuncCommand{
+	Name:  "shell",
+	Short: "Open a shell in a container",
+	OnInit: func(cmd *cobra.Command) {
+		cmd.PersistentFlags().StringSliceVar(&shellEntrypoint, "entrypoint", nil, "entrypoint to use")
+	},
+	OnSelectObject: func(c *callContext, name string) (*modTypeDef, error) {
+		if name == Container {
+			c.Select("id")
+			shellIsContainer = true
+			return &modTypeDef{Kind: dagger.Stringkind}, nil
+		}
+		return nil, nil
+	},
+	CheckReturnType: func(_ *callContext, _ *modTypeDef) error {
+		if !shellIsContainer {
+			return fmt.Errorf("shell can only be called on a container")
+		}
 
-func RunShell(
-	ctx context.Context,
-	engineClient *client.Client,
-	_ *dagger.Module,
-	cmd *cobra.Command,
-	dynamicCmdArgs []string,
-) (err error) {
-	rec := progrock.FromContext(ctx)
-	vtx := rec.Vertex(
-		digest.Digest("shell"),
-		"shell",
-	)
-	defer func() { vtx.Done(err) }()
+		// Even though these flags are global, we only check them just before query
+		// execution because you may want to debug an error during loading or for
+		// --help.
+		if silent || !(progress == "auto" && autoTTY || progress == "tty") {
+			return fmt.Errorf("running shell without the TUI is not supported")
+		}
+		if debug {
+			return fmt.Errorf("running shell with --debug is not supported")
+		}
 
-	cmd.SetOut(vtx.Stdout())
-	cmd.SetErr(vtx.Stderr())
+		return nil
+	},
+	OnResult: func(c *callContext, cmd *cobra.Command, returnType modTypeDef, result *any) error {
+		ctrID, ok := (*result).(string)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", result)
+		}
 
-	res, err := runQuery(ctx, engineClient, dynamicCmdArgs)
-	if err != nil {
-		return fmt.Errorf("failed to run query: %w", err)
-	}
+		ctx := cmd.Context()
+		ctr := c.e.Dagger().Container(dagger.ContainerOpts{
+			ID: dagger.ContainerID(ctrID),
+		})
 
-	var getCtrID func(x any) (string, error)
-	getCtrID = func(x any) (string, error) {
-		switch x := x.(type) {
-		case map[string]any:
-			for _, v := range x {
-				return getCtrID(v)
-			}
-			return "", fmt.Errorf("no container found")
-		case string:
-			return x, nil
-		default:
-			return "", fmt.Errorf("unexpected type %T", x)
+		if shellEntrypoint != nil {
+			ctr = ctr.WithExec(shellEntrypoint, dagger.ContainerWithExecOpts{
+				SkipEntrypoint: true,
```

Comment:
Think this makes sense (related to discussion the other day on some PR), but if the container doesn't have an entrypoint, then we'll end up dropping into a shell for the most recent `WithExec` on the container as set in the Function.

I feel like that becomes confusing to explain to users.

I _think_ the ideal might be:
1. (for follow-up, no need in this PR) support a flag on `dagger shell` that lets you override the command to run; use that if set
1. Otherwise, if entrypoint set, use that (already done here)
1. Finally, default to running `sh`

So, if you agree, then for this PR maybe we can just add that last case of defaulting to `sh` if no entrypoint. Then a follow-up can add the first priority of using the flag if set.

Reply:
Also, `DefaultArgs` fits into the ordering or priorities somewhere or another 🙂

Reply:
> 1. (for follow-up, no need in this PR) support a flag on dagger shell that lets you override the command to run; use that if set

That's already there with the `--entrypoint` flag, saved into the `shellEntrypoint` var when set.

Agree on all of that, here's the changes:
- If `--entrypoint` isn't set, check the container's entrypoint+defaultArgs;
- If no default command, run with `sh`;

Reply:
Done.